### PR TITLE
Speed Optimizations

### PR DESCRIPTION
**Pull request by metabench**
Please note - the code is not tidied up, there are lots of commented out console.log statements, I have left code which is yet to be deleted. I'm still working on the code, and plan to get more optimizations done. I didn't want to rearrange the code yet, but wanted to send what I had done once I had seen very high speed increases (about 334%, depending on how it's calculated).

-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~

Have replaced the use of many Int32Arrays with a single 'matrix'
Int32Array for each component. Created new versions of functions that
interact with that array.

Removed the 'buffer copy' code at the end by writing
getData_rgb_to_rgba_buffer, which writes directly to an RGBA buffer.

In the test I did, observed decoding time decreased from 3300ms to 760ms.

I also got it to calculate a few variables that get used multiple times, like XscaleX and YscaleY. I don't know if those other optimizations really sped it up, it was changing to the blocks_matrix that really made the difference, and then making it write the result directly to the RGBA node Buffer made a big difference too.

There are further optimizations that can yet be made, to do with better
use of typed arrays, and not allocating various variables multiple
times, but reusing space in typed arrays.

### CODE REVIEW

**Comment by metabench** [comment]
Hi @eugeneware (and anyone else interested),

Recently I've been busy working on a different project involving optimized JavaScript. I wound up making an OO structure to access a TypedArray, with functions to get and set values. I found it to be logical to program with, but ran a lot slower than I wanted. I then inlined various get and set function calls, and it ran really quickly. I'm still using the Class that contains a TypedArray inside, but refereing directly to its Float64Array.

I've given up on replacing local variables with typed array references. It's hard to keep track of, and I think the V8 optimizer is able to deal with local variables that are only used for one type efficiently. I had previously been thinking more in terms of the ASM JavaScript style (or my misunderstanding of it) but I'm writing the code for the V8 engine which does not focus on the ASM optimizations anyway.

I think this is the technique that should be used to optimize the Huffman table, if it is to get done. I'm not planning on doing it all that soon, maybe sometime though.

**Comment by metabench** [comment]
So I think this pull request is ready to merge, or at least I don't think I'll be able to send an improved version of the pull request algorithm-wise.

**Comment by eugeneware** [comment]
Hi @metabench - thanks for the update! Good luck with your TypedAdrray optimization for the future.

The current PR It doesn't rebase cleanly onto `HEAD`. If you could rebase and probably do a pass to remove any debug statements and commented out codes then we can merge and release.

We should probably bump the major version just in case too I'm thinking given the size of the change.

**Comment by metabench** [comment]
I'm much better at optimizing JavaScript than I am at using git and rebasing anything, Would it be OK for you to rebase the code that I have, or provide more detailed instructions about how I would do the rebase?

**Comment by neophob** [comment]
+1

**Comment by benwiley4000** [comment]
@metabench I also could be pushing my luck here, but do you still have access to the benchmarks you used to test these optimizations? 😃

**Comment by patrickhulce** [comment]
@benwiley4000 is your plan to try and minimize the diff here? It's quite hard to check what's going on with all the whitespace changes and feels quite risky in it's current state :)

Definitely worth re-benchmarking too to make sure there's a significant improvement in perf in modern JS engines.

**Comment by benwiley4000** [comment]
Right, I'd like to definitely test some benchmarks.

My hope was to first merge master into this branch, and then see if we can distill the changes down to what's most useful.

**Comment by metabench** [comment]
@benwiley4000 I've allowed edits from maintainers. I don't have these benchmarks immediately available. If I find them, I could add them to the project. Alternatively, I could redo them or describe them, they were not substantial pieces of code, just running an operation a few times (such as 10) and using timer measurements.

It's possible there have been optimizations on the master branch since my work, so I can't be sure that it would be >300%, as the measurements were taken at an earlier stage in the codebase. V8 was less optimized back in August 2014, so I'd expect different benchmark results.

**Comment by benwiley4000** [comment]
@metabench thanks! if you could at least describe them, that would be great 🙂

**Comment by metabench** [comment]
@benwiley4000 From my recollection, I found a large, high-resolution, and by my perception high-quality image of Mount Fuji. I used that as input to the decoder testers. I can't recall if one pass of decoding it was so slow normally (such as 3.5s) that I only needed one pass in the benchmarks.

I made two separate programs to perform the same operations using both the old and new code. Time differences were taken, and results used for comparison. I was careful only to take time measurements at the points in the code where the image processing was actually running, rather than during library initialization.

File: lib/decoder.js [hunk 1]
```diff
@@ -25,955 +25,1752 @@
...
-        }
-        k++;
-      }
-      if (i + 1 < length) {
-        // p here points to last code
-        code.push(q = {children: [], index: 0});
-        p.children[p.index] = q.children;
-        p = q;
-      }
+    "use strict";
```
Comment (XadillaX) [inline lib/decoder.js]: What happened to these lines?

Comment (metabench) [inline lib/decoder.js]: I don't know.  It does not look like I made any significant or useful changes to that part. Maybe they were some formatting changes that my IDE automatically made.

**Comment by patrickhulce** [comment]
thanks very much for your contributions @metabench ! since this has become quite stale, I'm going to go ahead and close for now.

if you're able to generate a cleaner diff again we'd be happy to reopen discussion around it :)

**Pull request closed without merging**